packages/@lwc/template-compiler/src/parser/index.ts

```diff
@@ -1142,14 +1141,14 @@ function isInIteration(ctx: ParserCtx): boolean {
     return !!ctx.findAncestor(ast.isForBlock);
 }
 
-function getForOfParent(ctx: ParserCtx, srcNode: ParentNode): ForOf | null {
-    return ctx.findAncestor(ast.isForOf, ({ current }) => !ast.isBaseElement(current), srcNode);
+function getForOfParent(ctx: ParserCtx): ForOf | null {
+    return ctx.findAncestor(ast.isForOf, ({ parent }) => parent && !ast.isBaseElement(parent));
 }
 
 function getForEachParent(ctx: ParserCtx): ForEach | null {
     return ctx.findAncestor(ast.isForEach, ({ parent }) => parent && !ast.isBaseElement(parent));
 }
 
-function isInIteratorElement(ctx: ParserCtx, parent: ParentNode): boolean {
-    return !!(getForOfParent(ctx, parent) || getForEachParent(ctx));
+function isInIteratorElement(ctx: ParserCtx): boolean {
+    return !!(getForOfParent(ctx) || getForEachParent(ctx));
```

Comment:
@jmsjtu @pmdartus This part was really tricky – we had slightly different logic for `getForOfParent` and `getForEachParent`. It was causing `getForOfParent` not to work with inline iterators, because I was ignoring the `key` in that case.

I guess this didn't matter much before, because these tests were just looking for an error case (keys should not reference the index, or iteration missing a key entirely).

In my fix, we don't care if the `for:each` or `iterator:x` is on the same node or the parent node. I'd appreciate if you took a look to make sure I didn't miss something.

Reply:
...Just confirmed that it's possible to use an `index` as a `key` if the `iterator` is inline. My PR fixes this.

I can open it as a separate PR if you prefer. This one is getting pretty complex.

Reply:
Opened a separate PR: https://github.com/salesforce/lwc/pull/2703